src/java/org/apache/cassandra/db/tries/SlicedTrie.java

```diff
@@ -0,0 +1,242 @@
+                if (newDepth == leftNextDepth && transition == leftNext)
+                {
+                    assert leftNext != ByteSource.END_OF_STREAM;
+                    leftNext = left.next();
+                    ++leftNextDepth;
+                    if (leftNext == ByteSource.END_OF_STREAM && includeLeft)
+                        state = State.INSIDE; // report the content on the left bound
+                }
+                else // otherwise we are beyond it
+                    state = State.INSIDE;
+            }
+
+            return checkRightBound(newDepth, transition);
+        }
+
+        private int markDone()
+        {
+            state = State.AFTER_RIGHT;
+            return -1;
+        }
+
+        private int checkRightBound(int newDepth, int transition)
+        {
+            // Cursor positions compare by depth descending and transition ascending.
+            if (newDepth > rightNextDepth)
+                return newDepth;
+            if (newDepth < rightNextDepth)
+                return markDone();
+            // newDepth == rightDepth
+            if (transition < rightNext)
+                return newDepth;
+            if (transition > rightNext)
+                return markDone();
+
+            // Following right bound
+            rightNext = right.next();
+            ++rightNextDepth;
+            if (rightNext == ByteSource.END_OF_STREAM && excludeRight)
+                return markDone();  // do not report any content on the right bound
+            return newDepth;
+        }
+
+        @Override
+        public int advanceMultiple(TransitionsReceiver receiver)
+        {
+            switch (state)
+            {
+                case BEFORE_LEFT:
+                    return advance();   // descend only one level to be able to compare cursors correctly
+                case INSIDE:
+                    int depth = source.depth();
+                    if (depth == rightNextDepth - 1)  // this is possible because right is already advanced;
+                        return advance();   // we need to check next byte against right boundary in this case
+                    int newDepth = source.advanceMultiple(receiver);
+                    if (newDepth > depth)
+                        return newDepth;    // successfully advanced
+                    // we ascended, check if we are still within boundaries
+                    return checkRightBound(newDepth, source.incomingTransition());
+                default:
+                    throw new AssertionError();
+            }
+        }
+
+        @Override
+        public int skipChildren()
+        {
+            assert (state != State.AFTER_RIGHT);
+
+            // We are either inside or following the left bound. In the latter case ascend takes us beyond it.
+            state = State.INSIDE;
+            return checkRightBound(source.skipChildren(), source.incomingTransition());
```

Comment:
nit: I'm having trouble finding a test that touches this codepath, at least in `org.apache.cassandra.db.tries`.

Reply:
Hmm, strange that I hadn't ported the intersection-on-intersection tests... added now.